avm/res/azure-stack-hci/cluster/main.bicep

```diff
@@ -0,0 +1,402 @@
+  adapter: string[]
+
+  @description('Required. The name of the network intent.')
+  name: string
+
+  @description('Required. Specify whether to override the adapter property. Use false by default.')
+  overrideAdapterProperty: bool
+
+  @description('Required. The adapter property overrides for the network intent.')
+  adapterPropertyOverrides: {
+    @description('Required. The jumboPacket configuration for the network adapters.')
+    jumboPacket: string
+
+    @description('Required. The networkDirect configuration for the network adapters.')
+    networkDirect: ('Enabled' | 'Disabled')
+
+    @description('Required. The networkDirectTechnology configuration for the network adapters.')
+    networkDirectTechnology: ('RoCEv2' | 'iWARP')
+  }
+
+  @description('Required. Specify whether to override the qosPolicy property. Use false by default.')
+  overrideQosPolicy: bool
+
+  @description('Required. The qosPolicy overrides for the network intent.')
+  qosPolicyOverrides: {
+    @description('Required. The bandwidthPercentage for the network intent. Recommend 50.')
+    bandwidthPercentage_SMB: string
+
+    @description('Required. Recommend 7.')
+    priorityValue8021Action_Cluster: string
+
+    @description('Required. Recommend 3.')
+    priorityValue8021Action_SMB: string
+  }
+
+  @description('Required. Specify whether to override the virtualSwitchConfiguration property. Use false by default.')
+  overrideVirtualSwitchConfiguration: bool
+
+  @description('Required. The virtualSwitchConfiguration overrides for the network intent.')
+  virtualSwitchConfigurationOverrides: {
+    @description('Required. The enableIov configuration for the network intent.')
+    enableIov: ('true' | 'false')
+
+    @description('Required. The loadBalancingAlgorithm configuration for the network intent.')
+    loadBalancingAlgorithm: ('Dynamic' | 'HyperVPort' | 'IPHash')
+  }
+
+  @description('Required. The traffic types for the network intent.')
+  trafficType: ('Compute' | 'Management' | 'Storage')[]
+}
+
+// define custom type for storage adapter IP info for 3-node switchless deployments
+@export()
+type storageAdapterIPInfoType = {
+  @description('Required. The HCI node name.')
+  physicalNode: string
+
+  @description('Required. The IPv4 address for the storage adapter.')
+  ipv4Address: string
+
+  @description('Required. The subnet mask for the storage adapter.')
+  subnetMask: string
+}
+
+// define custom type for storage network objects
+@export()
+type storageNetworksType = {
+  @description('Required. The name of the storage adapter.')
+  adapterName: string
+
+  @description('Required. The VLAN for the storage adapter.')
+  vlan: string
+
+  @description('Optional. The storage adapter IP information for 3-node switchless or manual config deployments.')
+  storageAdapterIPInfo: storageAdapterIPInfoType[]? // optional for switched deployments
+}
+@export()
+type storageNetworksArrayType = storageNetworksType[]
```

Comment:
I'd highly recommend to remove this snippet and instead change the parameter's implementation to
```bicep
@description('Required. An array of JSON objects that define the storage network configuration for the cluster. Each object should contain the adapterName, VLAN properties, and (optionally) IP configurations.')
param storageNetworks storageNetworksArrayType[]
```
As a heads up: There's a fair chance that there's also going to be a spec soon that asks users to also write UDTs as singular and non-nullable and instead configure these characteristics directly at the parameter to make it easier to understand how the template parameter(s) work

Reply:
By the way, this is also the reason all AVM-aligned common-types in the corresponding [utility module](https://github.com/Azure/bicep-registry-modules/blob/main/avm/utl/types/avm-common-types/main.bicep) are written in the way they are. We just published those recently.

Reply:
Updated type / params to match this pattern